commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+       * {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as
+       * {@link javax.cache.Cache#removeAll(Set)}.
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>It makes sense to expose global operation like this
+       *    instead of forcing users to iterate over the keys to lookup and call
+       *    get individually since Infinispan can do things more efficiently.
+       *    </li>
+       *    <li>Since this is a write-only operation, no entry attributes can be
+       *    queried, hence the only reasonable thing can be returned is Void.
+       *    </li>
+       *    <li>It still makes sense to return a CloseableIterator<Void> instead of
+       *    CompletableFuture<Void> in case the user wants to do something as each
+       *    operation completes, e.g. keep track of progress.
+       *    </li>
+       * </ul>
+       *
+       * @param keys the keys associated with each of the {@link WriteEntryView}
+       *             passed in the function callbacks
+       * @param f operation that the {@link WriteEntryView} associated with
+       *          one of the keys passed in
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalMany(Set<? extends K> keys, Consumer<WriteEntryView<V>> f);
+
+      /**
+       * Evaluate a write-only {@link Consumer} operation with the
+       * {@link WriteEntryView} of the value associated with the key, for all
+       * existing keys in functional map, and returns a
+       * {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as
+       * {@link javax.cache.Cache#removeAll()}.
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>In the original design, this method would have taken no
+       *    parameters and returned a CloseableIterator<WriteEntryView<V>>,
+       *    but doing so would have created encapsulation problems, since any
+       *    changes to the entry view would need to be committed, and that
+       *    would happen outside of the interceptor stack scope. To avoid this
+       *    problem, simply pass in a consumer with the desired operation and
+       *    then, the interceptor stack can control how the consumer gets
+       *    executed and make sure it's committed.</li>
+       * </ul>
+       *
+       * @param f operation that the {@link WriteEntryView} associated with
+       *          one of the keys passed in
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalAll(Consumer<WriteEntryView<V>> f);
```

Comment:
Yeah at least we have the choice :) - I personally love idempotent operations, but I understand they are not always possible.

If we don't enforce idempotency, I personally feel that having a second consumer would be a bit cleaner, however we are able to use BiConsumer still (passing the boolean), the problem is if we find we need another argument we will have to start using some unwieldly interfaces (TriConsumer etc.)